How many late days are you using for this assignment? zero

In [206]:
# Imports and pip installations (if needed)
%pip install pandas

import pandas as pd



Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


# Part 1: Load the dataset

In [207]:
# Load the given datasets
ckd_categorical_data = pd.read_csv("chronic_kidney_disease_categorical.csv")
ckd_numerical_data = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
display(ckd_categorical_data)
display(ckd_numerical_data)


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

The data in chronic kidney disease numerical dataset contains measurements, things that have units, such as age, blood pressure, and red blood cell
count. All the features in this set are numerical. The features in this set are: age, blood pressure, blood glucose random, blood urea, serum creatinine, sodium, potassium, hemoglobin, packed cell volume, white blood cell count, and red blood cell count. 3 numerical columns are age, blood pressure, and blood urea.

The data in the chronic kidney disease categorical dataset contains diagnoses and observations. All the features in this dataset are categorical some of which are ranked categorically with numbers, while others are strictly boolean in nature. The features in this set are albumin, sugar, red blood cell, pus cell, pus cell clumps, bacteria, hypertension, diabetes mellitus, coronary artery disease, appetite, pedal edema, and anemia. 3 categorical columns are appetite, hypertension, and anemia. There is also the only label in this dateset which is "class" which has the value of either ckd(chronic kidney disease) or not ckd(not chronic kidney disease).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [208]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
numerical_duplicates = ckd_numerical_data.duplicated().sum()
print("Total number of duplicates in numerical dataset rows:")
print(numerical_duplicates)

# Drop these duplicate rows
ckd_numerical_data_no_dups = ckd_numerical_data.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
categorical_duplicates = ckd_categorical_data.duplicated().sum()
print("Total number of duplicates in categorical dataset rows:")
print(categorical_duplicates)
ckd_categorical_data_no_dups = ckd_categorical_data.drop_duplicates()

display(ckd_numerical_data_no_dups)
display(ckd_categorical_data_no_dups)


Total number of duplicates in numerical dataset rows:
30
Total number of duplicates in categorical dataset rows:
30


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
396,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
398,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,good,no,yes,ckd
399,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:
I don't think all of the unique ids are present in both datasets. This is probably because not all of the unique ids that had categorical data collected also had numerical data collected for them and vice versa. In the combined table rows that are missing from one dataset will have NaN in all of the columns from the dataset they are missing from.

In [209]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_data = ckd_numerical_data_no_dups.merge(ckd_categorical_data_no_dups, on="unique_id", how="outer")

# #Print the combined dataset
combined_data


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,103570,24.0,80.0,125.0,NaN,NaN,136.0,3.5,15.4,43.0,5600.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,104675,65.0,80.0,215.0,133.0,2.5,NaN,NaN,13.2,41.0,NaN,NaN,2.0,1.0,normal,normal,present,notpresent,no,yes,no,good,no,no,ckd
3,105923,90.0,90.0,139.0,89.0,3.0,140.0,4.1,12.0,37.0,7900.0,3.9,1.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
4,107297,60.0,70.0,NaN,NaN,NaN,NaN,NaN,16.4,43.0,10800.0,5.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,11500.0,3.3,4.0,3.0,normal,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd
396,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,996412,70.0,60.0,120.0,40.0,0.5,140.0,4.6,16.0,43.0,4500.0,4.9,NaN,NaN,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
398,997629,35.0,NaN,99.0,30.0,0.5,135.0,4.9,15.4,48.0,5000.0,5.2,0.0,0.0,normal,normal,NaN,NaN,no,no,no,good,no,no,notckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [210]:
# Calculate the percentage of rows that contain atleast one missing value
num_rows = len(combined_data)
num_rows_with_missing_data = combined_data.isnull().any(axis=1).sum()

# Print %
percent = (num_rows_with_missing_data / num_rows) * 100
print(f"Percent of rows that contain at least one missing value: {percent:.2f}%")

# Drop these rows from the dataset
complete_combined_data = combined_data.dropna()

# Print the Dataset
complete_combined_data

Percent of rows that contain at least one missing value: 61.75%


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
6,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,8100.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
8,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,7100.0,5.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,6000.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
390,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,7900.0,3.9,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
391,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,10200.0,5.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
394,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,9800.0,6.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,11500.0,3.3,4.0,3.0,normal,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd


## Part 3.4: Sort the dataset according to the Labels

In [211]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_data = complete_combined_data.sort_values("Target").reset_index(drop= True)

# Print the dataset
sorted_data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,6700.0,3.9,4.0,0.0,normal,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,8800.0,3.4,3.0,4.0,normal,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,7900.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,7700.0,6.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,6700.0,6.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [212]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

data_with_dummies = pd.get_dummies(sorted_data, columns= categorical_columns, dtype= int)

# Print the dataset
data_with_dummies


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,6700.0,3.9,4.0,0.0,0,1,1,0,0,1,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,8800.0,3.4,3.0,4.0,0,1,0,1,1,0,1,0,0,1,0,1,1,0,0,1,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,7900.0,6.4,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,7700.0,6.3,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,6700.0,6.5,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer:

In the downstream modeling using this approach or using a single column for the binary variable is equivalent if each categorical variable could only take 2 values because the model could be able to interpret both approaches the same, since 1 can mean true and 0 can mean false similar to what the dummy variables would be representing. In addition to this it is only the same if it the model does not view these 0's and 1's as having some relationship to each other that does not exist, like a greater than or less than relationship.

When there are more than 2 values for a categorical variable and you map different numerical values to each of the values, such as no condiment: 0, ketchup: 1, mustard: 2, etc., then there is the chance that the model interprets these values as ordinal and certain values being better than others or being closer to each other in some way because of the numerical values representing them, even if that relationship does not exist. This would not be good because the model could make assumptions that do not remotely apply and heavily skew the models outputs.

Using our approach just marks a value as true or false or present depending on the value, removing the chance of the model assuming an ordinal relationship that does not exist.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [213]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

std_data = data_with_dummies[numerical_columns].std()
mean_data = data_with_dummies[numerical_columns].mean()

for col in numerical_columns:
    data_with_dummies = data_with_dummies[(data_with_dummies[col] > mean_data[col] - 3 * std_data[col]) & (data_with_dummies[col] < mean_data[col] + 3 * std_data[col])]

# Print the dataset
display(data_with_dummies)


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
5,397388,63.0,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,9800.0,4.2,2.0,2.0,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
7,438182,73.0,80.0,253.0,142.0,4.6,138.0,5.8,10.5,33.0,7200.0,4.3,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
8,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
9,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,7900.0,6.4,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,7700.0,6.3,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,6700.0,6.5,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [214]:
# Normalize the all Numerical Attributes in the dataset.

minv = data_with_dummies[numerical_columns].min()
maxv = data_with_dummies[numerical_columns].max()

data_with_dummies[numerical_columns] = (data_with_dummies[numerical_columns] - minv) / (maxv - minv)

# Print the dataset

data_with_dummies

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,0.395161,0.153846,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
5,397388,0.770270,1.00,0.901288,0.163399,0.345679,0.766667,0.171429,0.606557,0.631579,0.443548,0.410256,2.0,2.0,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
7,438182,0.905405,0.50,0.785408,0.862745,0.518519,0.600000,0.828571,0.401639,0.447368,0.233871,0.435897,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
8,474407,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,0.653226,0.333333,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
9,475200,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.485714,0.188525,0.263158,0.258065,0.205128,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,0.189189,0.25,0.227468,0.222222,0.074074,0.500000,0.257143,0.737705,0.736842,0.096774,0.564103,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,203466,0.216216,0.00,0.115880,0.052288,0.098765,0.600000,0.400000,0.647541,0.684211,0.290323,0.974359,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,482105,0.513514,0.00,0.107296,0.235294,0.012346,0.600000,0.371429,0.770492,0.894737,0.274194,0.948718,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,228849,0.702703,0.50,0.128755,0.261438,0.098765,0.666667,0.171429,0.688525,0.894737,0.193548,1.000000,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

The unique_id column is not appropriate for modeling and predictions. It provides no information that is relevant when building a model. It is just being used to identify rows in the processing steps and would have no influence on making predictions on ckd when training a model because it provides no information about the person that the data in the rest of the columns is about.

In [215]:
#Remove that column

cleaned_data = data_with_dummies.drop(columns=["unique_id"])

# Print the dataset
cleaned_data

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,0.395161,0.153846,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
5,0.770270,1.00,0.901288,0.163399,0.345679,0.766667,0.171429,0.606557,0.631579,0.443548,0.410256,2.0,2.0,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
7,0.905405,0.50,0.785408,0.862745,0.518519,0.600000,0.828571,0.401639,0.447368,0.233871,0.435897,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
8,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,0.653226,0.333333,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
9,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.485714,0.188525,0.263158,0.258065,0.205128,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,0.189189,0.25,0.227468,0.222222,0.074074,0.500000,0.257143,0.737705,0.736842,0.096774,0.564103,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,0.216216,0.00,0.115880,0.052288,0.098765,0.600000,0.400000,0.647541,0.684211,0.290323,0.974359,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,0.513514,0.00,0.107296,0.235294,0.012346,0.600000,0.371429,0.770492,0.894737,0.274194,0.948718,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,0.702703,0.50,0.128755,0.261438,0.098765,0.666667,0.171429,0.688525,0.894737,0.193548,1.000000,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [216]:
# Export the dataframe to a new csv file
cleaned_data.to_csv("cleaned_data.csv", index= False)

# Export the dataframe to a new json file
cleaned_data.to_json("cleaned_data.json", orient= "records", lines= True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

### Paste the markdown table 

| age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | al | su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----------|------|----------|----------|----------|----------|----------|---------|---------|-----------|----------|------|------|--------------|------------|-------------|-----------|----------------|--------------|----------------|------------|--------|---------|-------|--------|--------|---------|-------------|-------------|-------|--------|--------|---------|-------------|-----------------|
| 0.194805  | 0.75 | 0.158798  | 0.196078  | 0.160494  | 0.285714  | 0.206897  | 0.059406  | 0.000000  | 0.653226  | 0.257143  | 4.0   | 0.0   | 0     | 1     | 1     | 0     | 0     | 1     | 0     | 1     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 0     | 1     | 1     | 0     |
| 0.870130  | 1.00 | 0.965665  | 0.522876  | 0.641975  | 0.714286  | 0.000000  | 0.148515  | 0.225806  | 0.217742  | 0.057143  | 3.0   | 2.0   | 1     | 0     | 1     | 0     | 0     | 1     | 1     | 0     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 1     | 0     | 1     | 0     | 1     | 0     |
| 0.714286  | 0.50 | 0.442060  | 0.901961  | 0.432099  | 0.571429  | 0.793103  | 0.000000  | 0.032258  | 0.395161  | 0.057143  | 2.0   | 0.0   | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 1     | 0     |
| 0.649351  | 0.75 | 0.253219  | 0.633987  | 0.777778  | 0.457143  | 0.655172  | 0.138614  | 0.193548  | 0.169355  | 0.114286  | 2.0   | 0.0   | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 0     | 1     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     |
| 0.701299 | 0.75     | 0.150215 | 0.281046 | 0.234568 | 0.600000 | 0.793103 | 0.336634 | 0.322581 | 0.500000  | 0.314286  | 2.0       | 0.0       | 1         | 0         | 0         | 1         | 1         | 0         | 1         | 0         | 1         | 0         | 1         | 0         | 1         | 0         | 1         | 0         | 1         | 0         | 1         | 0         | 1         | 0         |
| 0.688312  | 0.50 | 1.000000  | 0.163399  | 0.111111  | 0.200000  | 0.206897  | 0.267327  | 0.387097  | 0.532258  | 0.371429  | 1.0   | 0.0   | 1     | 0     | 0     | 1     | 1     | 0     | 1     | 0     | 1     | 0     | 0     | 1     | 1     | 0     | 0     | 1     | 1     | 0     | 1     | 0     | 1     | 0     |
| 0.740260  | 1.00 | 0.901288  | 0.163399  | 0.345679  | 0.800000  | 0.206897  | 0.524752  | 0.548387  | 0.443548  | 0.342857  | 2.0   | 2.0   | 0     | 1     | 0     | 1     | 1     | 0     | 0     | 1     | 0     | 1     | 1     | 0     | 0     | 1     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     |
| 0.870130  | 0.50 | 0.785408  | 0.862745  | 0.518519  | 0.657143  | 1.000000  | 0.277228  | 0.322581  | 0.233871  | 0.371429  | 2.0   | 0.0   | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 0     | 1     | 0     | 1     | 0     | 1     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     |
| 0.753247  | 0.00 | 0.725322  | 0.313725  | 0.481481  | 0.628571  | 0.862069  | 0.178218  | 0.193548  | 0.258065  | 0.114286  | 4.0   | 1.0   | 1     | 0     | 1     | 0     | 1     | 0     | 0     | 1     | 0     | 1     | 0     | 1     | 1     | 0     | 0     | 1     | 0     | 1     | 1     | 0     | 1     | 0     |
| 0.636364  | 0.50 | 0.618026  | 0.411765  | 0.432099  | 0.628571  | 0.689655  | 0.316832  | 0.354839  | 0.250000  | 0.200000  | 3.0   | 1.0   | 0     | 1     | 1     | 0     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 1     | 0     | 1     | 0     | 0     | 1     | 1     | 0     | 1     | 0     |
| 0.818182  | 0.25 | 0.618026  | 0.562092  | 0.728395  | 0.142857  | 0.344828  | 0.168317  | 0.161290  | 0.580645  | 0.085714  | 4.0   | 3.0   | 0     | 1     | 1     | 0     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 0     | 1     | 1     | 0     | 0     | 1     | 0     | 1     | 1     | 0     |
| 0.844156  | 0.00 | 0.206009  | 0.751634  | 0.604938  | 0.600000  | 0.689655  | 0.366337  | 0.387097  | 0.879032  | 0.371429  | 4.0   | 0.0   | 0     | 1     | 0     | 1     | 1     | 0     | 1     | 0     | 0     | 1     | 0     | 1     | 1     | 0     | 0     | 1     | 0     | 1     | 1     | 0     | 1     | 0     |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions? 

When I was doing this 383 GPT often missed data, put random values, or just did not finish outputting the whole table. It took multiple prompts to get it to create the whole table. I had to have it first create the headers and then independently do very small chunks of rows at a time for it to create an accurate table.

You could mitigate data conversion hallucinations by breaking up the data given to the LLM into smaller chunks so it reduces the complexity for the model and it this makes it easier to compare the outputted data to the original data for errors since it is broken into smaller sections. In addition can give the LLM the data in structured data formats like csv, or json, instead of just raw text because it will be easier for it to process well structured data and also ask for the output in formatted data structure making it easier for the human reviewing the data to understand it. Finally making sure the prompts given to the LLM are clear and explicit so the model knows exactly what to do and no assumptions are left up to the model.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [217]:
# Assuming cleaned_data is your DataFrame
# Sample cleaned_data for demonstration (you should already have your own DataFrame)
# cleaned_data = pd.DataFrame(...)

# Define the mapping of abbreviations to full names
abbreviation_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rbcc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns in the DataFrame
cleaned_data.rename(columns=abbreviation_mapping, inplace=True)

# Display the updated DataFrame's columns
print(cleaned_data.columns)

display(cleaned_data)

Index(['age', 'blood pressure', 'blood glucose random', 'blood urea',
       'serum creatinine', 'sodium', 'potassium', 'hemoglobin',
       'packed cell volume', 'white blood cell count', 'red blood cell count',
       'albumin', 'sugar', 'rbc_abnormal', 'rbc_normal', 'pc_abnormal',
       'pc_normal', 'pcc_notpresent', 'pcc_present', 'ba_notpresent',
       'ba_present', 'htn_no', 'htn_yes', 'dm_no', 'dm_yes', 'cad_no',
       'cad_yes', 'appet_good', 'appet_poor', 'pe_no', 'pe_yes', 'ane_no',
       'ane_yes', 'Target_ckd', 'Target_notckd'],
      dtype='object')


,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,packed cell volume,white blood cell count,red blood cell count,albumin,sugar,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,0.395161,0.153846,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
5,0.770270,1.00,0.901288,0.163399,0.345679,0.766667,0.171429,0.606557,0.631579,0.443548,0.410256,2.0,2.0,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
7,0.905405,0.50,0.785408,0.862745,0.518519,0.600000,0.828571,0.401639,0.447368,0.233871,0.435897,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
8,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,0.653226,0.333333,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
9,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.485714,0.188525,0.263158,0.258065,0.205128,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,0.189189,0.25,0.227468,0.222222,0.074074,0.500000,0.257143,0.737705,0.736842,0.096774,0.564103,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,0.216216,0.00,0.115880,0.052288,0.098765,0.600000,0.400000,0.647541,0.684211,0.290323,0.974359,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,0.513514,0.00,0.107296,0.235294,0.012346,0.600000,0.371429,0.770492,0.894737,0.274194,0.948718,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,0.702703,0.50,0.128755,0.261438,0.098765,0.666667,0.171429,0.688525,0.894737,0.193548,1.000000,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [218]:
# Converted SQL to Pandas code

# Assuming your DataFrame is named cleaned_data

# Group by 'Target_ckd' and count occurrences
result = cleaned_data.groupby('Target_ckd').size().reset_index(name='count')

# Display the result
print(result)

   Target_ckd  count
0           0    111
1           1     23


*Describe what the above code does here*

The code, groupby('Target_ckd'), groups the data in the cleaned_data data frame by the values in the 'Target_ckd' column and then the size() function counts the number of rows in each group, returning a series. Then reset_index(name="count") turns the series from size() back into a DataFrame and assigns the count to a new column named count. In the end a new DataFrame is created with the values in the Target_ckd column and the number of occurrences of each.
